test/functorch/test_control_flow.py

```diff
@@ -1471,44 +1726,860 @@ def fct_pointwise(x, y):
         z = torch.randn(3, 2, 2, device=device, requires_grad=True)
         inp = {"i": x, "j": ([y], [{"o": z}])}
 
-        with self.assertRaisesRegex(Exception, r"."):
+        with self.assertRaisesRegex(
+            # Should be: RuntimeError,
+            # r"The number of leaves of the pytree of the output of the operator
+            # needs to match the lenght of the pytree of the input",
+            torch._dynamo.exc.Unsupported,
+            "Observed exception.*",
+        ):
             result = associative_scan(fct_wrong_pytree, inp, 0, combine_mode="generic")
 
+        result = associative_scan(
+            fct_pointwise, inp, 0, combine_mode=combine_mode, reverse=reverse
+        )
+        expected_result = _fake_associative_scan(fct_pointwise, inp, 0, reverse=reverse)
+        self.assertEqual(result, expected_result)
+
+    @requires_cuda
+    @parametrize("reverse", [False, True])
+    @parametrize("device", [torch.device("cpu"), torch.device("cuda")])
+    def test_scan_complex_pytree(self, reverse, device):
+        # Init and input have same pytree
```

Comment:
init and input doesn't have to have the same pytree. We should remove this restriction.

Reply:
Done. The only restrictions now are between the init and the carry. They need to have the same dtype, shape and device